src/goto-symex/symex_goto.cpp

```diff
@@ -126,12 +148,71 @@ void goto_symext::symex_goto(statet &state)
     state_pc=goto_target;
   }
 
+  // Goto is conditional, so figure out which branch to take before
+  // checking whether to do path exploration or model-checking, and before
+  // adjusting guards.
+  switch(static_cast<optionst::goto_priorityt>(
+    options.get_unsigned_int_option("goto-priority")))
+  {
+  case optionst::goto_priorityt::JUMP:
+  {
+    goto_programt::const_targett tmp = new_state_pc;
+    new_state_pc = state_pc;
+    state_pc = tmp;
+    forward = !forward;
+    break;
+  }
+  case optionst::goto_priorityt::NEXT:
+    break;
+  }
+
+  // Normally the next instruction to execute would be state_pc and we save
+  // new_state_pc for later. But if we're executing from a saved state, then
+  // new_state_pc should be the state that we saved from earlier, so let's
+  // execute that instead.
+  if(state.has_saved_target)
+  {
+    INVARIANT(
+      new_state_pc == state.saved_target,
+      "Tried to explore the other path of a branch, but the next "
+      "instruction along that path is not the same as the instruction "
+      "that we saved at the branch point. Saved instruction is " +
+        state.saved_target->code.pretty() +
+        "\nwe were intending "
+        "to explore " +
+        new_state_pc->code.pretty() +
+        "\nthe "
+        "instruction we think we saw on a previous path exploration is " +
+        state_pc->code.pretty());
+    goto_programt::const_targett tmp = new_state_pc;
+    new_state_pc = state_pc;
+    state_pc = tmp;
+    log.debug() << "Resuming from '" << state_pc->code.source_location() << "'"
+                << log.eom;
+  }
+  else if(options.get_bool_option("paths"))
+  {
+    // At this point, `state_pc` is the instruction we should execute
+    // immediately, and `new_state_pc` is the instruction that we should execute
+    // later (after we've finished exploring this branch). For path-based
+    // exploration, save `new_state_pc` to the saved state that we will resume
+    // executing from, so that goto_symex::symex_goto() knows that we've already
+    // explored the branch starting from `state_pc` when it is later called at
+    // this branch.
+    branch_pointt branch_point(target, state);
+    branch_point.state.saved_target = new_state_pc;
+    branch_point.state.has_saved_target = true;
+    branch_point.state.saved_target_is_backwards = forward;
```

Comment:
I think this is correct, but add a comment clarifying the counterintuitive "backward = forward" statement here